### PR TITLE
Add custom dataset and dataloader tutorial for C++

### PR DESCRIPTION
**Pull request by dhpollack**
I was struggling with creating a custom dataset and dataloader for the
C++ frontend partially because I couldn't find a good minimal example
and ran into a few gotchas when trying to model something after the
MNIST dataset.  This tutorial might help others avoid the pitfalls that
I ran into.

Signed-off-by: David Pollack <david@da3.net>

### CODE REVIEW

**Comment by boulabiar** [comment]
I found this example
https://github.com/pytorch/examples/blob/master/cpp/custom-dataset/custom-dataset.cpp

File: advanced_source/cpp_dataset.rst [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+  If you need ``CMAKE_PREFIX_PATH`` to be something other than the 
+  location to the ``libtorch`` directory, you can use the environmental 
+  variable ``Torch_DIR`` as an alternative.  To do so, you export 
+  it in your shell as follows:
+
+.. code-block:: shell
+
+  export Torch_DIR=/full/path/to/libtorch/share/cmake/Torch
+  # or
+  export Torch_DIR=$(realpath rel/path/to/libtorch/share/cmake/Torch)
```
Comment (glaringlee) [inline advanced_source/cpp_dataset.rst:78]: `rel` means relative path right? if so, can we use full word to make this clear?

File: advanced_source/cpp_dataset.rst [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+Ok, if we want to iterate through our dataset a single example at a time then
+we are done.  Of course, we probably want to process our data in minibatches.
+As in the python frontend we also have a ``Dataloader`` utility class in the
+C++ frontend.  However, in the python frontend there is a parameter in the
+``Dataloader`` class called ``collate_fn``.  The default collation function
+basically converts a list of basic python types, numpy arrays, or torch tensors
+stacks them into a minibatch.  One normally doesn't need to write a custom
+collation function except for special cases such as sequence data when you need
+to pad a sequence to the length of the longest sequence in the batch.  In the
+C++ frontent, the equivalent to the collation function are mapping transforms.
```
Comment (glaringlee) [inline advanced_source/cpp_dataset.rst:280]: frontent->frontend

File: advanced_source/cpp_dataset.rst [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+    }
+
+    return 1;
+  }
+
+Now let's build and run the code.
+
+.. code-block:: shell
+
+  export Torch_DIR=$(realpath rel/path/to/libtorch/share/cmake/Torch)
```
Comment (glaringlee) [inline advanced_source/cpp_dataset.rst:398]: same here. rel -> relative

**Review by glaringlee — COMMENTED** [review]
@dhpollack Thanks for contributing this tutorial. This generally looks good to me. I slightly commented it.

This tutorial have 2 major parts, one is how to create a customized dataset/dataloader workflow, the other one is how to build it with libtorch. Can you put two reference link into this tutorial as well? We have a code example for creating dataset/dataloader and we have a documentation for build libtorch with customized code, but we have no tutorial to put all of these together, I think this tutorial is a good place to have all of these reference.
code example: https://github.com/pytorch/examples/blob/master/cpp/custom-dataset/custom-dataset.cpp
libtorch build document: https://pytorch.org/cppdocs/installing.html

**Comment by glaringlee** [comment]
There are some CI tests failures, but it seems not related to this PR. Let's see whether CI tests is recovered after this PR is updated.

**Comment by dhpollack** [comment]
@glaringlee @jspisak I need to test this with the newest version of pytorch as I created this over a year ago and haven't tried it out with the newest versions of libtorch.  I am on vacation this week, but I can try it this weekend / Monday when I get back.

**Comment by dhpollack** [comment]
@glaringlee I tested this with libtorch 1.8 and gcc 10.2 and it works on my system.  Good to go unless you have some other comments.

**Comment by glaringlee** [comment]
@dhpollack This is good to me, there is a CI failure, we will merge this after clear the CI failure.

**Comment by glaringlee** [comment]
@brianjo 
`ci/circleci: pytorch_tutorial_windows_pr_build_worker_0` is failed due to a python module missing.  Is this an issue?

**Review by glaringlee — APPROVED** [review]
@jspisak @brianjo This looks good to me. There is a CI failure in `ci/circleci: pytorch_tutorial_windows_pr_build_worker_0`, but should not be related to this PR in my opinion.

**Comment by dhpollack** [comment]
> @brianjo
> `ci/circleci: pytorch_tutorial_windows_pr_build_worker_0` is failed due to a python module missing. Is this an issue?

Yea, librosa is an audio module. Probably related to one of the audio tutorials. Has nothing to do with this PR.

**Comment by brianjo** [comment]
Rebasing this some fixes I added to master.

**Comment by jspisak** [comment]
@guyang3532 - looks like some Windows failures are occuring - can you take a look?

**Comment by brianjo** [comment]
> @guyang3532 - looks like some Windows failures are occuring - can you take a look?

I think this just needs a rebase. I did that. Should fix the issues.

**Comment by brianjo** [comment]
Sorry, looks like there is a Windows issue on this one.


________________________________

**Comment by dhpollack** [comment]
stale

**Pull request closed without merging**